### ISSUE TITLE
Screen reader accessibility review hoping for more improvements in the future

### ISSUE THREAD

**Issue by pvagner** [movim/movim#1581]
Screen reader accessibility work is one of the keypoints introduced in the latest movim release and I have took a stab at it from the real screen reader users perspective.

When navigating movim UI with a screen reader running parts of the UI seem to be confusing or would benefit from clearer labelling.

I am blind so I can't imagine visual appearance of the parts I am describing. My descriptions are based off of screen reader presentation of the UI.

I very much appreciate screen reader accessibility is mentioned in social media in the blogposts and I'd be happy if it continues to be improved further.

## navigation

* list with single item
    * My profile
        Link is focusable and there is profile name displayed alongside it. I assume the profile name is more important than the fact it's the profile so it should be communicated more prominently to the assistive technologies such as screen readers. Menu item role is not appropriate here at all as the element with role="menuitem" is not keyboard focusable and the focus is not managed by the app in this part.
    * single item list
        * chat_bubble
            And again not so easy to discover text 1 messages.
    * another single item list
        * Add
            With the text Add a new space
    * List with three items
        * newsmode
            Icon name followed by the actual text News
        * explore
            with the actual text Explore
    * note_stack_add
            Publish something
    * list with two items
        * search
            Search
        * notifications
            Notifications

### Notes and questions

To me it looks lists and list items are used instead of containers here. Other than serving as containers do these lists have other important functionality? i.e. would the appearance become hindered if they will be changed to div elements?

Links are focusable elements here but the whole list items are used as iconized clickable elements perhaps resembling desktop style controls. For accessibility it would perhaps be nicer if the whole items would become focusable with appropriate aria roles. As accessible name calculation as implemented within the web browsers is doing correct thing here plus it will avoid code duplication.

As for grouping I can't make up my mind if they are just multiple buttons and aria toolbar would be a nice accessibility paradigm for improving its accessibility or if this part is about to mimic a menu bar like appearance where aria-menu would be welcome implementation.

For both these approaches keyboard focus handling would have to be implemented in javascript.

If java script dependency should be minimized then it should perhaps stay like a simple navigation with links, and the list containers should perhaps be reconsidered.

## Main

The actual main element receives the focus which I think is unnecessary

* list with 7 items
    * chevron_right
        Explore. Of course chevron_right should become aria-hidden and the same thing to the items above applies related to naming and keyboard focus handling.
    * and 6 more news items
    Grouping inside aclearly labelled container would perhaps be a good idea to better know these are news. Or perhaps a different term should be used I don't yet know the app very well.
* list with single item
    * share
        Share. Again list container + aria-hidden + labelling +
* post_add
    this time it's a button with icon that should get aria-hidden but with helpfull tooltip saying "Write a new post". I assume new post into a pubsub feed.
* list with 3 items
    * links all, contacts and communities.
        Contacts and Community have tooltips, all don't. I don't know the app very well to tell if this needs improvements. If it should resemble a desktop style app perhaps a list of tabs with tab pannel would be  better. (again proper keyboard focus handling needed once this is going to be considered)
* News feet content
    I don't yet follow published blogs like this.
    Finally there is a heading level 4 with an empty line and a link. Using heading an empty line like this is not really appropriate here. Proper CSS theming should be used instead. The link has an explore icon that should be aria-hidden

## Messaging

Messaging / chatting is the first thing XMPP client is good for so let's try that with movim.

Find a link saying chat_bubble in the navigation landmark inside first single item list and explore how the content under main landmark changes.

This time the main element does not receive the keyboard focus.

* First there is a list with common contacts. It's designed as a list. Its first item is kind of heading. For accessibility it would be better if the heading would be outside the list and the list would be labelled with that heading content. Unfortunately each item in the list has a link with the text "chat" inside and the contact details such as status, message, name are surrounding text parts. The list is not interactive so the only way on how to effectively navigate in this part of the UI is by using screen reader features rather than the app features. While using tab or shift+tab to navigate each contact is presented as "chat link" with a screen readers. Again the part I have discussed earlier that the whole item should be focusable / actionable should be considered here. Preferably this should become an interactive list that only takes single tabstop and arrow keys are used to navigate its contents.
* Start a new conversation button
    this is fine.
* list with single item
    * Find some new contacts chevron_right
        No keyboard access within this part + chevron_right should become aria-hidden.
* list with 8 items
    No keyboard access within this part. Contacts are listed but from the keyboard there is nothing actionable here.
* menu
    * list with single item
        * Start a new conversation button
            It would have been better if the menu role was not used. Button does not belong into a menu.
    * labelled list stories with single item
        * web_stories
            aria-hidden + list container + labelling... It is really looking as if the authors were experimenting with various techniques waiting for comments like this trying to find out what works the best.
    * list of contacts
        The same thing that I was saying about common contacts also applies here. If it will ever become an interactive list that won't be correct having list inside a menu.
    * list with single item
The item is saying messages all contacts and is not actionable so I am really confused here. Perhaps it's meant as a tab list.
    * list with active conversations
        Nick name of the chat partner is listed as the actionable link. message preview and the app name have to be read with the screen reader features.
    * list of joined group chats
        There is another list nested inside list. In that nested list there is a link named chat_bubble_outline with the tooltip with the room topic. I'd prefer if this was interactive list.

## The actual conversation

After opening a conversation the chat text entry receives the focus and I can type in my message.

From the top of the conversation window we do have these elements

* list with single item
    * single item with multiple controls: arrow_back, clickable unlabelled image (presumably contact avatar), manage_search, phone, videocam, more_vert, contact's nick name, contacts status
        just a clickable elements, difficult to activate with a screen reader running, not focusable with the keyboard lacking clear labelling. Although some of the controls have tooltips because of poor semantics these are not reported at all.
* list with multiple items
    * icons person, report (both with accompanying text), and a text clear history, then separator, then block and close.
        not actionable, not keyboard focusable, difficult to handle with a screen reader. Perhaps this should be made into aria toolbar with buttons as roving tab index.
* list with messages
    time and date is listed twice, avatar is not actionable and is okay to become aria-hidden. Context icons such as add a reaction and reply are not actionable and focusable.
* clock_arrow_down button and expand_more button
    focusable and actionable but poorly labelled
* some nested lists with links and other controls
    There are links with icons such as sticker, mic, emoji_emotions, add_circle. These are actionable and poorly labelled.
    icons camera_alt, gesture, attach_file have the text alongside, icons should be aria-hidden, and when clearly labelled controls should be made focusable and keyboard navigable.

I choosed to activate the mic link however it revealed controls that are not keyboard focusable, are not clearly labelled and are very difficult to use with a screen reader.

## Summary

The fact need for screen reader accessibility is already recognized within the project is something I'm proud of. I admire all the developers and contributors, I appreciate all the work and all the efforts that have been and are going to be put into this effort however there is much more work ahead as compared with the work that has been done up to now.
For day to day use screen reader accessibility of movim is not yet viable to screen reader users.

Feel free to ask for more ideas like this if you find them usefull.

Thanks and greetings

Peter

**Comment by thegregariousdragon** [comment]
Completely agree with Peter here... As a [NVDA](https://nvaccess.org) user, the experience is definitely hit or miss with the interface. As someone who has their own server, there's a few things I like, but that could use some imporvement:

1. I can access ad-hoc commands, but the need to travel to the top of the interface, then inbvoke NVDA's focus mode to enter a dialog is not ideal. It's usable, meaning I can perform commands, but frankly the experience can be improved. I'll admit that I'm not as good explaining what needs to be done to improve things as I don't know the ins and outs of NVDA as well as some others might.
2. As Peter pointed out, the chat interface can be improved. One suggestion I'd make here is to have new messages in the chat be exposed to screen readers so they can be read out automatically so the chatting experience can feel more natural. Right now, a NVDA user has to leave focus mode, return to browse mode and move up the interface to read any messages, which isn't ideal.
3. Adding a profile picture is a bit of a pain with NVDA, as you have to use keyboard commands to right click on the edit link, which NVDA renders in an awkward fashion, and once you figure that bit out, then it's another trip to the top of the screen, enter focus mode to enter the dialog, then a return to browse mode to find the button when in the dialog to open the file picker. Also not ideal.

I could go on, but I'll need to interact with the interface some more before I can add additional UX feedback.
---
TGDragon

**Comment by edhelas** [comment]
Thanks for the feedback, will work on it when I have time :)

**Comment by pvagner** [comment]
Is there something I might try helping with in order to start improving screen reader accessibility bit by bit?

Would this be a good opportunity to try to find answers to at least some of the high level questions I have raised in my attempt at an screen reader accessibility review back in may this year?

For example those list items and other containers consisting of parent container, then icon, then actionable control such as a link or a button, then some complementary text. Click handler is configured on the parent, keyboard accessibility is wired by the fact activating the focusable element inside click on the parent element. For keyboard only sighted users this is working fine but for screen reader users it's problematic as we need to have some more context exposed rather than the action name such as *Chat* on the interactive control inside. In order to improve this I could imagine creating an aria-label for the interactive control inside however doing that we have to duplicate the browser's accessible name calculation our-selves almost everywhere also potentially leading to a lot of duplicate code.
Changing parent clickable elements to become focus targets is something I am afraid touching my-self as it may also affect visual appearance.

Please, Would you be keen on looking to this in a foreseable future @edhelas and if we can find a compromise I would be happy to continue with gradually working on small improvements adapting the chosen strategy to the different parts of the app.

Or is this something you can't devote your time this year?

I see this can't be made accessible over night but my goal would be to start moving slowly and perhaps after a few weeks / months we could have something to be proud of then.

**Comment by Menelion** [comment]
I quickly looked at the code. It's a huge bunch of PHP templates, which is good syntax-wise (everything is absolutely clear), but accessibility-wise… it's a lot of work to be done. And by "a lot" I mean **a lot**. Many `<i>` items without any role, most of those I have no idea about, so even I took this task (which is really time-consuming), I'd need to be in permanent contact with core devs simply to know what is what. So @pvagner, yes, it's not over night and even not a week of work.
